### PR TITLE
Replacing some of the UDF

### PR DESCRIPTION
**Pull request by TMaile**

### CODE REVIEW

File: lib/buildingsync/model_articulation/building_section.rb [hunk 1]
```diff
@@ -193,10 +193,44 @@ def read_construction_types(section_element, ns)
       end
     end
 
+    def add_principal_hvac_type(buildingSection)
+      # code here
+      buildingSections = buildingSection.parent
```
Comment (corymosiman12) [inline lib/buildingsync/model_articulation/building_section.rb]: please use snake_case for variable names

File: lib/buildingsync/model_articulation/hvac_system.rb [hunk 2]
```diff
@@ -69,6 +70,41 @@ def get_primary_hvac_system_type
       return nil
     end
 
+    # adding the principal hvac system type to the hvac systems, overwrite existing values or create new elements if none are present
+    def add_primary_system_type(id, principal_hvac_type)
```
Comment (corymosiman12) [inline lib/buildingsync/model_articulation/hvac_system.rb]: refactor as `add_principal_hvac_system_type` to match updates for BSync > 2.1.

File: lib/buildingsync/model_articulation/hvac_system.rb [hunk 3]
```diff
@@ -69,6 +70,41 @@ def get_primary_hvac_system_type
...
+            end
+          end
+        end
+        if hvac_system.nil? and @systems.elements["#{ns}:HVACSystem"].size = 1
+          hvac_system = @systems.elements["#{ns}:HVACSystem"][0]
+        end
+      end
+
+      if hvac_system.elements["#{ns}:PrimaryHVACSystemType"].nil?
+        primary_hvac_system_type = REXML::Element.new("#{ns}:PrimaryHVACSystemType")
```
Comment (corymosiman12) [inline lib/buildingsync/model_articulation/hvac_system.rb]: rename to `PrincipalHVACSystemType`

File: lib/buildingsync/model_articulation/spatial_element.rb [hunk 4]
```diff
@@ -281,18 +281,18 @@ def add_element_in_xml_file(building_element, ns, field_name, field_value)
...
-      add_element_in_xml_file(xml_element, ns, 'BarDivisionMethod', @bar_division_method)
-      add_element_in_xml_file(xml_element, ns, 'FractionArea', @fraction_area)
-      add_element_in_xml_file(xml_element, ns, 'SpaceTypesFloorArea', @space_types_floor_area)
-      add_element_in_xml_file(xml_element, ns, 'ConditionedFloorAreaHeatedOnly', @conditioned_floor_area_heated_only)
-      add_element_in_xml_file(xml_element, ns, 'ConditionedFloorAreaCooledOnly', @conditioned_floor_area_cooled_only)
-      add_element_in_xml_file(xml_element, ns, 'ConditionedFloorAreaHeatedCooled', @conditioned_floor_area_heated_cooled)
-      add_element_in_xml_file(xml_element, ns, 'ConditionedBelowGradeFloorArea', @conditioned_below_grade_floor_area)
-      add_element_in_xml_file(xml_element, ns, 'CustomConditionedAboveGradeFloorArea', @custom_conditioned_above_grade_floor_area)
-      add_element_in_xml_file(xml_element, ns, 'CustomConditionedBelowGradeFloorArea', @custom_conditioned_below_grade_floor_area)
+      add_user_defined_field_to_xml_file(xml_element, ns, 'TotalFloorArea', @total_floor_area)
```
Comment (corymosiman12) [inline lib/buildingsync/model_articulation/spatial_element.rb]: Is `TotalFloorArea` the same as the gross floor area?  If so, this should not be added as a UDF.  Same goes for the other floor areas below:
```xml

              <auc:FloorAreas>
                <auc:FloorArea>
                  <auc:FloorAreaType>Cooled only</auc:FloorAreaType>
                  <auc:FloorAreaValue>0.0</auc:FloorAreaValue>
                </auc:FloorArea>
                <auc:FloorArea>
                  <auc:FloorAreaType>Gross</auc:FloorAreaType>
                  <auc:FloorAreaValue>49390</auc:FloorAreaValue>
                </auc:FloorArea>
                <auc:FloorArea>
                  <auc:FloorAreaType>Heated and Cooled</auc:FloorAreaType>
                  <auc:FloorAreaValue>49261</auc:FloorAreaValue>
                </auc:FloorArea>
                <auc:FloorArea>
                  <auc:FloorAreaType>Heated only</auc:FloorAreaType>
                  <auc:FloorAreaValue>129</auc:FloorAreaValue>
                </auc:FloorArea>
              </auc:FloorAreas>
```

File: lib/buildingsync/model_articulation/spatial_element.rb [hunk 5]
```diff
@@ -281,18 +281,18 @@ def add_element_in_xml_file(building_element, ns, field_name, field_value)
...
+      add_user_defined_field_to_xml_file(xml_element, ns, 'BuildingType', @bldg_type)
+      add_user_defined_field_to_xml_file(xml_element, ns, 'SystemType', @system_type)
+      add_user_defined_field_to_xml_file(xml_element, ns, 'BarDivisionMethod', @bar_division_method)
+      add_user_defined_field_to_xml_file(xml_element, ns, 'FractionArea', @fraction_area)
+      add_user_defined_field_to_xml_file(xml_element, ns, 'SpaceTypesFloorArea', @space_types_floor_area)
+      add_user_defined_field_to_xml_file(xml_element, ns, 'ConditionedFloorAreaHeatedOnly', @conditioned_floor_area_heated_only)
+      add_user_defined_field_to_xml_file(xml_element, ns, 'ConditionedFloorAreaCooledOnly', @conditioned_floor_area_cooled_only)
+      add_user_defined_field_to_xml_file(xml_element, ns, 'ConditionedFloorAreaHeatedCooled', @conditioned_floor_area_heated_cooled)
+      add_user_defined_field_to_xml_file(xml_element, ns, 'ConditionedBelowGradeFloorArea', @conditioned_below_grade_floor_area)
+      add_user_defined_field_to_xml_file(xml_element, ns, 'CustomConditionedAboveGradeFloorArea', @custom_conditioned_above_grade_floor_area)
```
Comment (corymosiman12) [inline lib/buildingsync/model_articulation/spatial_element.rb]: What does custom refer to here?

File: lib/buildingsync/model_articulation/spatial_element.rb [hunk 6]
```diff
@@ -281,18 +281,18 @@ def add_element_in_xml_file(building_element, ns, field_name, field_value)
...
-      add_element_in_xml_file(xml_element, ns, 'ConditionedFloorAreaHeatedCooled', @conditioned_floor_area_heated_cooled)
-      add_element_in_xml_file(xml_element, ns, 'ConditionedBelowGradeFloorArea', @conditioned_below_grade_floor_area)
-      add_element_in_xml_file(xml_element, ns, 'CustomConditionedAboveGradeFloorArea', @custom_conditioned_above_grade_floor_area)
-      add_element_in_xml_file(xml_element, ns, 'CustomConditionedBelowGradeFloorArea', @custom_conditioned_below_grade_floor_area)
+      add_user_defined_field_to_xml_file(xml_element, ns, 'TotalFloorArea', @total_floor_area)
+      add_user_defined_field_to_xml_file(xml_element, ns, 'BuildingType', @bldg_type)
+      add_user_defined_field_to_xml_file(xml_element, ns, 'SystemType', @system_type)
+      add_user_defined_field_to_xml_file(xml_element, ns, 'BarDivisionMethod', @bar_division_method)
+      add_user_defined_field_to_xml_file(xml_element, ns, 'FractionArea', @fraction_area)
+      add_user_defined_field_to_xml_file(xml_element, ns, 'SpaceTypesFloorArea', @space_types_floor_area)
```
Comment (corymosiman12) [inline lib/buildingsync/model_articulation/spatial_element.rb]: What does `SpaceTypesFloorArea` and `FractionArea` refer to?

**Review by corymosiman12 — CHANGES_REQUESTED** [review]
Please also run rubucop on this.  I think it should catch some things like `nonSnakeCase` for variables and others

**Review by corymosiman12 — CHANGES_REQUESTED** [review]
Please also run rubucop on this.  I think it should catch some things like `nonSnakeCase` for variables and others

File: lib/buildingsync/model_articulation/building_section.rb (hunk 1 above)
Comment (TMaile) [reply lib/buildingsync/model_articulation/building_section.rb]: fixed in a26368e

File: lib/buildingsync/model_articulation/hvac_system.rb (hunk 2 above)
Comment (TMaile) [reply lib/buildingsync/model_articulation/hvac_system.rb]: addressed in 6884510

File: lib/buildingsync/model_articulation/hvac_system.rb (hunk 3 above)
Comment (TMaile) [reply lib/buildingsync/model_articulation/hvac_system.rb]: addressed in 6884510

Comment (TMaile) [reply lib/buildingsync/model_articulation/hvac_system.rb]: and 882a7f1

File: lib/buildingsync/model_articulation/spatial_element.rb (hunk 4 above)
Comment (TMaile) [reply lib/buildingsync/model_articulation/spatial_element.rb]: @mrigeshgupta please look at: https://github.com/BuildingSync/BuildingSync-gem/blob/7117b72a443496d9d74e93461b597e85664f3bbe/lib/buildingsync/model_articulation/spatial_element.rb#L54 to write the properties to the same XML elements as we read them from. 
We want to write those only if they are not null.

File: lib/buildingsync/model_articulation/spatial_element.rb (hunk 5 above)
Comment (TMaile) [reply lib/buildingsync/model_articulation/spatial_element.rb]: @corymosiman12 Custom refers to the FloorAreaType = 'Custom', basically a custom string defining the floor area. We could also skip this here in the write process if you think this is too esoteric.

File: lib/buildingsync/model_articulation/spatial_element.rb (hunk 6 above)
Comment (TMaile) [reply lib/buildingsync/model_articulation/spatial_element.rb]: @corymosiman12 
fraction_area defined the fraction of the section/building/element of the total area
space_types_floor_area is actually a hash that contains the floor area of each space type

**Comment by TMaile** [comment]
> Please also run rubucop on this. I think it should catch some things like `nonSnakeCase` for variables and others

@mrigeshgupta can you please do that

**Pull request merged**